README.md
Please see the official documentation at [docs.strongloop.com](http://docs.strongloop.com/display/LB/Installing+the+Oracle+connector)


CHANGES.md
2016-11-29, Version 3.0.0
=========================

 * Default oracledb module to 'oracledb' (Raymond Feng)

 * Improve user install experience (#15) (Rashmi Hunt)

 * Update URLs in CONTRIBUTING.md (#12) (Ryan Graham)

 * Switch to oracledb (Raymond Feng)


2016-05-04, Version 1.4.6
=========================

 * Fix typo and catch error (Raymond Feng)


2016-05-04, Version 1.4.5
=========================

 * Handle the package if the main module is loopback-connector-oracle (Raymond Feng)


2016-05-03, Version 1.4.4
=========================

 * update/insert copyright notices (Ryan Graham)

 * relicense as MIT only (Ryan Graham)


2016-04-05, Version 1.4.3
=========================

 * Fix the installer for npm3 (Raymond Feng)


2016-01-24, Version 1.4.2
=========================

 * Add a better message to report not found (Raymond Feng)

 * Refer to licenses with a link (Sam Roberts)

 * Use strongloop conventions for licensing (Sam Roberts)


2015-07-29, Version 1.4.1
=========================

 * Allow env var to override config.oracleUrl (Raymond Feng)


2015-03-24, Version 1.4.0
=========================

 * Detects node versions (Raymond Feng)


2015-01-06, Version 1.3.0
=========================

 * Update deps (Raymond Feng)

 * Fix bad CLA URL in CONTRIBUTING.md (Ryan Graham)

 * Use /bin/sh, to avoid reliance on script exec bit (Sam Roberts)

 * Add contribution guidelines (Ryan Graham)

 * Fix link to doc (Raymond Feng)

 * Update MacOSX install message (superkhau)

 * Update Linux install message (superkhau)


2014-05-22, Version 1.2.1
=========================

 * Update deps (Raymond Feng)

 * Fix the url protocol for proxy detection (Raymond Feng)


2014-02-28, Version 1.2.0
=========================

 * Generate strong-oracle.rc under home directory (Raymond Feng)

 * Use strong-oracle module (Raymond Feng)


2014-02-18, Version 1.1.3
=========================

 * Bump version (Raymond Feng)

 * Reformat code (Raymond Feng)

 * Update to dual MIT/StrongLoop license (Raymond Feng)

 * Retrieve the proxy from npm config (Raymond Feng)


2014-02-14, Version 1.1.2
=========================

 * Add a link to docs (Raymond Feng)

 * Allow the HTTP_PROXY env var to configure http proxy (Raymond Feng)

 * Replace old README with link to docs. (Rand McKinney)

 * Remove blanket (Raymond Feng)

 * Check the existency of local binary (Raymond Feng)

 * Ignore local installation errors (Raymond Feng)

 * Fix windows space issue and update dependencies (Raymond Feng)

 * Fix SLA-517 - Terminated two unterminated strings. (Edmond Meinfelder)

 * Update docs (Raymond Feng)


2013-09-11, Version strongloopsuite-1.0.0-5
===========================================



2013-09-11, Version strongloopsuite-1.0.0-4
===========================================



2013-09-11, Version strongloopsuite-1.0.0-3
===========================================

 * Add keywords to package.json (Raymond Feng)


2013-09-10, Version strongloopsuite-1.0.0-2
===========================================

 * Finalize package.json for sls-1.0.0 (Raymond Feng)


2013-09-04, Version strongloopsuite-1.0.0-1
===========================================



2013-09-04, Version strongloopsuite-1.0.0-0
===========================================

 * First release!


pkginstall.js
// Copyright IBM Corp. 2013,2016. All Rights Reserved.
// Node module: loopback-oracle-installer
// This file is licensed under the MIT License.
// License text available at https://opensource.org/licenses/MIT

var path = require('path');
var download = require('./lib/download');

var url = process.argv[2] || process.env.LOOPBACK_ORACLE_URL;
var dest = process.argv[3];
var version = null;
var oracledbModule = 'oracledb';

if (typeof dest === 'undefined' || dest == null) {
  var connector = false;
  var oracleConnectorPath;

  //find the directory of oracle-cocnnector-installer
  var installerIndex = require.resolve('.');
  installerPath = path.dirname(installerIndex);

  try {
    //loopback-connector-oracle can be resolved only in apiconnect project env where loopback-connector-oracle
    //is a pier module of loopback-oracle-installer
    var connector = require.resolve('loopback-connector-oracle');
    oracleConnectorPath = path.dirname(connector);
    connector = true;
  } catch (err) {
    //check if parent module is loopback-connector-oracle for loopback-oracler-installer by traversing to /loopback-oracler-installer/../../loopback-connector-oracle
    var parentDir = path.dirname(path.join(installerPath, '..'));
    if (parentDir.includes('loopback-connector-oracle')) {
      //this is a dev env usecase where developer has checked out loopback-connector-oracle repo for dev/test.
      //In this case loopback-connector-oracle exists as a parent of loopback-oracler-installer.
      //traverse to loopback-connector-oracle installation location
      oracleConnectorPath = parentDir;
      connector = true;
    } else {
      //this is a dev env usecase where developer has checked out loopback-oracler-installer repo for dev/test.
      connector = false;
    }
  }
  //calculate the location to install 'oracledb' node module
  if (connector) { //loopback-connector-oracle is either a pier or a parent module of loopback-oracle-installer.
      //In apiconnect project env and loopback-connector-oracle dev env, oracledb will be
      //installed under /loopback-connector-oracle/node_modules
      dest = path.normalize(path.join(oracleConnectorPath, 'node_modules'));
  } else { //loopback-connector-oracle is neither a pier nor a parent module of loopback-oracle-installer.
      //This is just a loopback-oracle-installer dev env. Install oracledb under loopback-oracle-installer/node_modules.
      dest = path.normalize(path.join(installerPath, 'node_modules'));
  }

  // Find the package.json for loopback-connector-oracle
  if (oracleConnectorPath) {
    var pkg = require(path.join(oracleConnectorPath, 'package.json'));
    if (pkg.config && pkg.config.oracleUrl) {
      // Allow env var to override config.oracleUrl
      url = url || pkg.config.oracleUrl;
      version = pkg.config.oracleVersion;
      oracledbModule = pkg.config.driverModule || oracledbModule;
    }

    var installValues = "Installation values -> oracledb location: " + dest + ' Prepackaged tarball url: ' + url + "  oracleversion: " + version + " oracledbmodule: " + oracledbModule + '\n';
    console.log(installValues);
  }
}

var downloadRequired = false;
try {
  if (!process.env.npm_config_force) {
    var found = require.resolve('oracledb');
    console.log('** Skipping oracledb as it already exists at %s)', path.dirname(found));
    console.log('** To force download and re-installation, run npm install -f.');
  } else {
    downloadRequired = true;
  }
} catch (e) {
  downloadRequired = true;
}

if (downloadRequired) {
  // First download the archive
  download(url, version, dest, oracledbModule, function(err, result) {
    if (err) {
      process.exit(1);
    }
  });
}




lib/fstream-patch.js
// Copyright IBM Corp. 2013,2014. All Rights Reserved.
// Node module: loopback-oracle-installer
// This file is licensed under the MIT License.
// License text available at https://opensource.org/licenses/MIT

/**
 * Monkey-patch fstream.FileWriter to boost the performance of node-tar Extract which always
 * write in 512 byte chunks
 */
var Writer = require('fstream').Writer;
var FileWriter = require('fstream').FileWriter;
var bufStream = require('buffered-stream');

var fs = require("graceful-fs"), EOF = {};

FileWriter.prototype._create = function () {
  var me = this;
  if (me._stream) {
    return;
  }

  var so = {};
  if (me.props.flags) {
    so.flags = me.props.flags;
  }
  so.mode = Writer.filemode;
  if (me._old && me._old.blksize) {
    so.bufferSize = me._old.blksize;
  }

  me._fstream = fs.createWriteStream(me._path, so);
  me._stream = bufStream(1024 * 1024); // Set up a buffered stream to aggregate small writes
  me._stream.pipe(me._fstream);

  me._fstream.on("open", function (fd) {
    // console.error("FW open", me._buffer, me._path)
    me.ready = true;
    me._buffer.forEach(function (c) {
      if (c === EOF) {
        me._stream.end();
      }
      else {
        me._stream.write(c);
      }
    });
    me.emit("ready");
    // give this a kick just in case it needs it.
    me.emit("drain");
  });

  me._fstream.on("drain", function () {
    me.emit("drain");
  });

  me._fstream.on("close", function () {
    // console.error("\n\nFW Stream Close", me._path, me.size)
    me._finish();
  });
};



lib/download.js
// Copyright IBM Corp. 2013,2016. All Rights Reserved.
// Node module: loopback-oracle-installer
// This file is licensed under the MIT License.
// License text available at https://opensource.org/licenses/MIT

/**
 * This is a workaround to node-tar which writes in chunks of 512 bytes by monkey-patching the fs api
 */
require('./fstream-patch');

var zlib = require('zlib');
var tar = require('tar');
var fs = require('fs');
var fse = require('fs-extra');
var homedir = require('os-homedir');
var path = require('path');
var url = require('url');
var request = require('request');
var spawn = require('cross-spawn');
var info = require('./detect.js');

/*
 * Get the http/https proxy for the give url string from the npm configuration
 * The npm configuration properties are passed into the script as
 * process.env.npm_config_<npm_property_name>
 */
function getProxyFor(urlStr) {

  var urlObj = url.parse(urlStr);

  if (urlObj.protocol === 'http:') {
    // https://www.npmjs.org/doc/misc/npm-config.html#proxy
    return process.env.npm_config_proxy;
  }
  if (urlObj.protocol === 'https:') {
    // https://www.npmjs.org/doc/misc/npm-config.html#https-proxy
    return process.env.npm_config_https_proxy || process.env.npm_config_proxy;
  }
  return null;
}

module.exports = function (url, version, dest, oracledbModule, cb) {
  var abi = '';
  if (info.nodeVersion !== undefined) {
    abi = '-abi' + info.nodeVersion;
  }
  var targz = 'loopback-oracle-' + info.platform + '-' + info.arch + abi +
    '-' + (version || info.version) + '.tar.gz';


  if (!url) {
    //for development testing purpose
    url = path.join(dest, '../../loopback-oracle-builder/build', info.platform, targz);
  } else {
    url = url + '/' + targz;
  }

  console.log('Downloading/Extracting ' + url + '\n');
  if (!dest) {
    dest = './node_modules';
  }
  var gunzip = zlib.createGunzip({chunkSize: 1024 * 1024});
  var src = null;
  var local = false;
  if (url.indexOf('http://') === 0 || url.indexOf('https://') === 0) {
    var options = {};
    var proxyUrl = getProxyFor(url);
    if (proxyUrl) {
      console.log('HTTP proxy: ' + proxyUrl);
      options = {
        proxy: proxyUrl
      };
    }
    src = request.get(url, options);
  } else {
    if (fs.existsSync(url)) {
      src = fs.createReadStream(url);
      local = true;
    } else {
      checkPrereqsAndInstallOracledb(url, dest, oracledbModule, cb);
    }
  }
  var fileFound = true;
  if (src) {
    if (!local) {
      src.on('error', function (error) {
        checkPrereqsAndInstallOracledb(url, dest, oracledbModule, cb);
      });
      src.on('response', function (response) {
        //pre-packaged tarball not found. Attempt to install oracledb module if user has installed oracle pre-requisites offline.
        if (response.statusCode === 404) {
          checkPrereqsAndInstallOracledb(url, dest, oracledbModule, cb);
        } else { //pre-packaged tarball found. Extract the tarball.
            extractTarball(src, url, dest, gunzip, cb);
        }
      });
    } else { //tar found in filesystem under oracle-builder
      extractTarball(src, url, dest, gunzip, cb);
    }
  }
};

function extractTarball(src, url, dest, gunzip, cb) {
  src.pipe(gunzip).pipe(tar.Extract({path: dest}))
      .on('end', function () {
        console.log(url + ' is now extracted to ' + dest);
        var err;
        var home = homedir();
        var icDir = path.join(home, 'oracle-instant-client');
        if (!fs.existsSync(icDir)) {
          try {
            fse.copySync(path.join(dest, 'instantclient'), icDir,
                {preserveTimestamps: true});
            console.log('Oracle instant client is installed to %s', icDir);
          } catch (e) {
            err = e;
            console.error(e);
          }
        } else {
          console.warn('Oracle instant client found at %s', icDir);
        }
        cb && cb(err);
      }).on('error', function (error) {
    cb && cb(error);
  });
}
function checkPrereqsAndInstallOracledb(url, dest, oracledbModule, cb) {
  console.log('Pre-packaged file ' +  url  + ' doesn\'t exist.' + ' Checking if user has installed pre-requisites offline as per instructions in https://github.com/oracle/node-oracledb/blob/master/INSTALL.md' + '\n');
  //can not find the oracle instance client prepackaged tarball.  Check if user has installed offline oracle instance client
  //libraries and necessary pre-reqs as per instructions https://github.com/oracle/node-oracledb/blob/master/INSTALL.md
  if (process.env.OCI_LIB_DIR != null && process.env.OCI_INC_DIR != null) { // check the oracle pre-req variables are set
    //save current working dir
    var currentDir = process.cwd();
    process.chdir(dest);
    console.log('Attempting to install ' + oracledbModule + ' node module under ' + dest + '.' + ' If the installation fails with errors, please install pre-requisites for oracledb offline as per https://github.com/oracle/node-oracledb/blob/master/INSTALL.md'
        + ' and retry installing loopback-oracle-connector.' + '\n');
    // Spawn NPM asynchronously
    var child = spawn('npm', ['install', oracledbModule, '-depth', '0'], { stdio: 'inherit' });
    //restore back to the orginal working directory. (not sure if this step is needed)
    process.chdir(currentDir);
    console.log('Found Oracledb pre-requisites...');
    console.log('Installed Oracledb node module successfully.');
  } else {
    var notFound = 'Oracledb pre-requisites not found.' + ' User needs to install pre-requisites for oracledb offline as per https://github.com/oracle/node-oracledb/blob/master/INSTALL.md'
        + ' and retry installing loopback-oracle-connector.';
    console.error(notFound);
    cb && cb(notFound);
  }
}

if (require.main === module) {
  module.exports(process.argv[2], process.argv[3]);
}
